File: skills/agent-harness/scripts/materialize_private_template.py
#!/usr/bin/env python3
"""Copy a JSON template with a valid absolute file URI for its installed schema."""
from __future__ import annotations

import argparse
import hashlib
import json
import os
from pathlib import Path
import stat
import tempfile
from typing import Any

import check_authorization


class MaterializeError(ValueError):
    pass


def _regular_json(path: Path, label: str) -> dict[str, Any]:
    if not path.is_absolute() or path.is_symlink() or not path.is_file():
        raise MaterializeError(f"{label} must be an absolute regular non-symlink file")
    try:
        value = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as error:
        raise MaterializeError(f"{label} is not readable JSON") from error
    if not isinstance(value, dict):
        raise MaterializeError(f"{label} must contain a JSON object")
    return value


def materialize(
    template_path: Path,
    schema_path: Path,
    output_path: Path,
    *,
    replace: bool = False,
) -> dict[str, str]:
    template = _regular_json(template_path, "template")
    schema = _regular_json(schema_path, "schema")
    if not output_path.is_absolute():
        raise MaterializeError("output must be an absolute path")
    parent = output_path.parent
    if not parent.is_dir() or parent.is_symlink():
        raise MaterializeError("output parent must exist and must not be a symlink")
    if output_path.is_symlink() or (output_path.exists() and not output_path.is_file()):
        raise MaterializeError("output must be a regular non-symlink file")
    if output_path.exists() and not replace:
        raise MaterializeError("output already exists; pass --replace for an explicit update")

    canonical_schema = schema_path.resolve(strict=True)
    document = dict(template)
    document["$schema"] = canonical_schema.as_uri()
    if "contract_schema_id" in document:
        schema_id = schema.get("$id")
        if not isinstance(schema_id, str) or not schema_id:
            raise MaterializeError("schema must provide a stable non-empty $id")
        document["contract_schema_id"] = schema_id
    if "contract_schema_sha256" in document:
        document["contract_schema_sha256"] = (
            "sha256:" + hashlib.sha256(canonical_schema.read_bytes()).hexdigest()
        )
    validation_errors = check_authorization._schema_errors(document, schema)
    if validation_errors:
        raise MaterializeError(
            "materialized document failed its installed schema: "
            + "; ".join(sorted(set(validation_errors)))
        )

    descriptor, temporary = tempfile.mkstemp(
        prefix=f".{output_path.name}.", dir=os.fspath(parent)
    )
    try:
        os.fchmod(descriptor, stat.S_IRUSR | stat.S_IWUSR)
        with os.fdopen(descriptor, "w", encoding="utf-8") as handle:
            json.dump(document, handle, indent=2, sort_keys=False)
            handle.write("\n")
            handle.flush()
            os.fsync(handle.fileno())
        os.replace(temporary, output_path)
    finally:
        if os.path.exists(temporary):
            os.unlink(temporary)
    return {
        "output_path": str(output_path.resolve(strict=True)),
        "schema_uri": canonical_schema.as_uri(),
        "contract_schema_id": str(schema.get("$id", "")),
        "contract_schema_sha256": "sha256:"
        + hashlib.sha256(canonical_schema.read_bytes()).hexdigest(),
    }


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--template", type=Path, required=True)
    parser.add_argument("--schema", type=Path, required=True)
    parser.add_argument("--output", type=Path, required=True)
    parser.add_argument("--replace", action="store_true")
    arguments = parser.parse_args()
    try:
        result = materialize(
            arguments.template,
            arguments.schema,
            arguments.output,
            replace=arguments.replace,
        )
    except MaterializeError as error:
        print(json.dumps({"status": "blocked", "reason": str(error)}, sort_keys=True))
        return 2
    print(json.dumps({"status": "ok", **result}, sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/agent-harness/verification/Sources/AppleVerificationCore/MaterializePrivateTemplate.swift
import Darwin
import Foundation

public struct MaterializeError: Error, Equatable, CustomStringConvertible {
  public let message: String
  public init(_ message: String) { self.message = message }
  public var description: String { message }
}

public enum MaterializePrivateTemplate {
  private static func regularJSON(_ path: URL, label: String) throws -> [String: Any] {
    guard path.path.hasPrefix("/"), !isSymlink(path), isRegular(path) else {
      throw MaterializeError("\(label) must be an absolute regular non-symlink file")
    }
    do { return try HarnessRuntime.object(path) } catch {
      throw MaterializeError("\(label) is not readable JSON")
    }
  }

  private static func isSymlink(_ url: URL) -> Bool {
    (try? url.resourceValues(forKeys: [.isSymbolicLinkKey]).isSymbolicLink) == true
  }
  private static func isRegular(_ url: URL) -> Bool {
    (try? url.resourceValues(forKeys: [.isRegularFileKey]).isRegularFile) == true
  }

  public static func materialize(
    templatePath: URL, schemaPath: URL, outputPath: URL, replace: Bool = false
  ) throws -> [String: Any] {
    let template = try regularJSON(templatePath, label: "template")
    let schema = try regularJSON(schemaPath, label: "schema")
    guard outputPath.path.hasPrefix("/") else {
      throw MaterializeError("output must be an absolute path")
    }
    let parent = outputPath.deletingLastPathComponent()
    guard (try? parent.resourceValues(forKeys: [.isDirectoryKey]).isDirectory) == true,
      !isSymlink(parent)
    else { throw MaterializeError("output parent must exist and must not be a symlink") }
    let exists = FileManager.default.fileExists(atPath: outputPath.path)
    guard !isSymlink(outputPath), !exists || isRegular(outputPath) else {
      throw MaterializeError("output must be a regular non-symlink file")
    }
    guard !exists || replace else {
      throw MaterializeError("output already exists; pass --replace for an explicit update")
    }
    let canonicalSchema = schemaPath.resolvingSymlinksInPath().standardizedFileURL
    var document = template
    document["$schema"] = canonicalSchema.absoluteString
    if document["contract_schema_id"] != nil {
      guard let schemaID = schema["$id"] as? String, !schemaID.isEmpty else {
        throw MaterializeError("schema must provide a stable non-empty $id")
      }
      document["contract_schema_id"] = schemaID
    }
    let schemaDigest = "sha256:" + (try HarnessRuntime.sha256File(canonicalSchema))
    if document["contract_schema_sha256"] != nil {
      document["contract_schema_sha256"] = schemaDigest
    }
    let errors = JSONSchemaValidator.errors(
      instance: document, schema: schema, path: "$", root: nil)
    guard errors.isEmpty else {
      throw MaterializeError(
        "materialized document failed its installed schema: "
          + Array(Set(errors)).sorted().joined(separator: "; "))
    }
    do {
      try HarnessRuntime.atomicWriteJSON(document, to: outputPath)
      guard chmod(outputPath.path, S_IRUSR | S_IWUSR) == 0 else { throw POSIXError(.EACCES) }
    } catch { throw MaterializeError("output cannot be written atomically") }
    return [
      "output_path": outputPath.resolvingSymlinksInPath().path,
      "schema_uri": canonicalSchema.absoluteString,
      "contract_schema_id": schema["$id"] as? String ?? "", "contract_schema_sha256": schemaDigest,
    ]
  }

  public static func run(arguments: [String], context: RuntimeContext) throws -> Int32 {
    var options: [String: String] = [:]
    var replace = false
    var index = 0
    while index < arguments.count {
      let key = arguments[index]
      if key == "--replace" {
        guard !replace else { throw MaterializeError("invalid arguments") }
        replace = true
        index += 1
        continue
      }
      guard ["--template", "--schema", "--output"].contains(key), options[key] == nil,
        index + 1 < arguments.count
      else { throw MaterializeError("invalid arguments") }
      options[key] = arguments[index + 1]
      index += 2
    }
    guard let template = options["--template"], let schema = options["--schema"],
      let output = options["--output"]
    else { throw MaterializeError("invalid arguments") }
    do {
      let result = try materialize(
        templatePath: URL(fileURLWithPath: template), schemaPath: URL(fileURLWithPath: schema),
        outputPath: URL(fileURLWithPath: output), replace: replace)
      var response = result
      response["status"] = "ok"
      FileHandle.standardOutput.write(
        try HarnessRuntime.canonicalJSON(response, ensureASCII: true) + Data([0x0a]))
      return 0
    } catch let error as MaterializeError {
      let response: [String: Any] = ["status": "blocked", "reason": error.message]
      FileHandle.standardOutput.write(
        try HarnessRuntime.canonicalJSON(response, ensureASCII: true) + Data([0x0a]))
      return 2
    }
  }
}
